nibabel/arrayproxy.py

```diff
@@ -155,12 +226,34 @@ def inter(self):
     def is_proxy(self):
         return True
 
+    @contextmanager
+    def _get_fileobj(self):
+        """Create and return a new ``ImageOpener``, or return an existing one.
+
+        The specific behaviour depends on the value of the ``keep_file_open``
+        flag that was passed to ``__init__``.
+
+
+        Yields
+        ------
+        ImageOpener
+            A newly created ``ImageOpener`` instance, or an existing one,
+            which provides access to the file.
+        """
+        if self._keep_file_open:
+            if not hasattr(self, '_opener'):
+                self._opener = ImageOpener(self.file_like)
+            yield self._opener
+        else:
+            with ImageOpener(self.file_like) as opener:
+                yield opener
+
     def get_unscaled(self):
-        ''' Read of data from file
+        """ Read of data from file
 
         This is an optional part of the proxy API
-        '''
-        with ImageOpener(self.file_like) as fileobj:
+        """
+        with self._get_fileobj() as fileobj, self._lock:
```

Comment:
I was thinking `array_from_file` would take a lock and block its critical section. Do you have a sense over what kind of difference in slowdown could result from blocking the whole call instead? If it's trivial, then this is fine.

Reply:
I think, given that the `array_from_file` function reads an entire image, so where the lock occurs is not going to make much of a difference to performance. In fact, I am wondering whether this locking should occur in the `ArrayProxy` at all, or if it should be left to the responsibility of calling code.

This situation is a bit different to the locking needed in the `read_segments` function, which can't easily be managed by calling code  ... unless we delegate responsibilty entirely to the calling code, and allow a `Lock` to be passed into `ArrayProxy.__init__`?